### PR TITLE
Changes on HoareOptimizer

### LINKED ISSUE #4981: HoareOptimizer does not handle CZ correctly
### Information

- **Qiskit Terra version**: master
- **Python version**: 3.7.8
- **Operating system**: MacOS

### What is the current behavior?

Consider the following program, where `circuit_in` is optimized by `HoareOptimizer` pass as `circuit_out` :
```python
from numpy import pi
from qiskit import *
from qiskit.converters import *
from qiskit.transpiler.passes import HoareOptimizer
from qiskit.transpiler.passmanager import PassManager

circuit_in = QuantumCircuit(2)
circuit_in.h(0)
circuit_in.x(1)
circuit_in.cz(0, 1)
circuit_in.h(0)
circuit_in.measure_all()

def print_circuit_and_counts(circuit):
    print(circuit)
    backend = BasicAer.get_backend('qasm_simulator')    
    print(execute(circuit, backend=backend, optimization_level=0).result().get_counts())

print_circuit_and_counts(circuit_in)

pm = PassManager()
pm.append(HoareOptimizer(size=10))
circuit_out = pm.run(circuit_in)

print_circuit_and_counts(circuit_out)
```
```
        ┌───┐   ┌───┐ ░ ┌─┐   
   q_0: ┤ H ├─■─┤ H ├─░─┤M├───
        ├───┤ │ └───┘ ░ └╥┘┌─┐
   q_1: ┤ X ├─■───────░──╫─┤M├
        └───┘         ░  ║ └╥┘
meas: 2/═════════════════╩══╩═
                         0  1 
{'11': 1024}
              ░ ┌─┐   
   q_0: ──────░─┤M├───
        ┌───┐ ░ └╥┘┌─┐
   q_1: ┤ X ├─░──╫─┤M├
        └───┘ ░  ║ └╥┘
meas: 2/═════════╩══╩═
                 0  1 
{'10': 1024}
```

The optimization fully removes the `H-CZ-H` sequence wrongly, breaking the semantics equivalence of the transformation.

### What is the expected behavior?

I'm not fully sure if the original paper deals with phase (maybe the root of the problem). From what I remember, it seems to me that the qubit `q_1` falls into a "trivial condition" and the CZ should be replaced by a Z gate in `q_0`.

Notice that `circuit_in.cz(1, 0)` (instead of `circuit_in.cz(0, 1)` does not break the counting, since `HoareOptimizer` does not detect any optimization, which is also unexpected. Maybe the problem is related to #4980?

### PR DESCRIPTION
**Pull request by georgios-ts**
### Summary
Changes on ` HoareOptimizer `. Resolves #4981.

### Details and comments
This PR introduces the following changes on `HoareOptimizer`.

- Fixes triviality check condition for `CZGate`, `CU1Gate`,  `MCU1Gate`.  
These gates act trivially (up to global phase) iff the clause  or  together with previous conditions is unsatisfiable.

- Removes 'always trivial' condition from `CRZGate`.

- Catches `CircuitError` for instructions with no inverse.
For example, the following code would previously raise an error.
```
qc = QuantumCircuit(1)
qc.x(0)
qc.measure_all()

pm = PassManager()
pm.append(HoareOptimizer(size=10))
circ = pm.run(qc)
```

- Adds another optmization step by removing the control from a Controlled - U (and applies U directly on target qubits) iff all control qubits are in definite |1> state.

### CODE REVIEW

**Comment by 1ucian0** [comment]
While reviewing this PR with @javabster, we noticed that the `qiskit/transpiler/passes/optimization/_gate_extension.py` docstring was a bit twisted. @georgios-ts,  hope you dont mind we  pushed a change in this branch ( 126b955fa6d715e09b3738b331d481862a0486a9 ) with, what we think, is a better docstring.

File: test/python/transpiler/test_hoare_opt.py [hunk 1]
```diff
@@ -293,6 +293,45 @@ def test_targetsuccessive_identity_advanced_removal(self):
 
         self.assertEqual(result, circuit_to_dag(expected))
 
+    def test_control_removal(self):
+        """ Should replace CX by X
+            and CZ by Z. """
```
Comment (1ucian0) [inline test/python/transpiler/test_hoare_opt.py:298]:
```suggestion
        """ Should replace CX by X and CZ by Z. """
```

File: qiskit/transpiler/passes/optimization/hoare_opt.py [hunk 2]
```diff
@@ -138,6 +141,55 @@ def _test_gate(self, gate, ctrl_ones, trgtvar):
...
+        Args:
+            gate (Gate): gate to inspect
+            ctrlvar (list(BoolRef)): z3 variables corresponding to latest state
+                                     of control qubits
+            trgtvar (list(BoolRef)): z3 variables corresponding to latest state
+                                     of target qubits
+        Returns:
+            bool: if controlled gate can be replaced.
+            DAGCircuit: with U applied to the target qubits.
+            List: with indices of target qubits.
```
Comment (1ucian0) [inline qiskit/transpiler/passes/optimization/hoare_opt.py:156]:
```suggestion
        Returns:
            Tuple(bool, DAGCircuit, List)::
              * bool:if controlled gate can be replaced.
              * DAGCircuit: with U applied to the target qubits.
              * List: with indices of target qubits.
```

**Review by ajavadia — APPROVED** [review]
thanks for the fix, please also add a bugfix release note

**Review by ajavadia — APPROVED** [review]

**Pull request merged**